gax/src/main/java/com/google/api/gax/batching/BlockingSemaphore.java

```diff
@@ -43,31 +46,100 @@ private static void checkNotNegative(long l) {
   BlockingSemaphore(long permits) {
     checkNotNegative(permits);
     this.currentPermits = permits;
+    this.limit = permits;
   }
 
-  public synchronized void release(long permits) {
+  public void release(long permits) {
     checkNotNegative(permits);
-
-    currentPermits += permits;
-    notifyAll();
+    synchronized (updateLock) {
```

Comment:
I'm not exactly sure what is going on here, but based on the names and semantics of these locks, it looks like the job [ReadWriteLock](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReadWriteLock.html) was created for, doesn't it?. If yes, please just use the single read/write lock. In general, two locks going together seem suspicious and most importantly once there there is more than one lock it gets very easy to get deadlocked.

Reply:
Yeah sorry, 2 locks are not necessary. Revert it back to the original implementation.